client/DESIGN.md

```diff
@@ -0,0 +1,292 @@
+
+The HTTP request for write should be similar to the following:
+
+    HTTP/1.1 200 Okay
+    Value-TxClock: 1421236393816753
+
+However, it is also possible the DB server will be unable to complete the write because another client has written new values in the meantime.  In this case, the HTTP write resposne will be of the following form:
+
+    HTTP/1.1 412 Precondition Failed 
+
+In this case, the the client should retry the transaction with a smaller value for the min-age header (e.g., min-age=0) to ensure the values it receives from the DB are not stale.  
+
+<a name="Transaction"></a>
+# Transaction
+
+- Parameters: 
+    1. cache: Cache (required)
+    2. read\_timestamp: int (read\_timestamp=None)
+    3. max\_age: int (max\_age=None)
+    4. no\_cache: bool (no\_cache=False)
+- Public Methods: 
+    1. read(table: string, key: string, read\_timestamp=currentTime, max\_age=None, no\_cache=False) returns value (JSON) or None
+    2. write(table: string, key: string, value: JSON) returns None (always succeeds)
+    3. delete(table: string, key: string) returns None
+    4. commit() returns None, throws StaleException
+- Internal Objects: 
+    1. min\_Rt: int (The minimum read time the Transaction has seen so far, based on values read)
+    2. max\_Vt: int (The maximum value time the Transaction has seen so far, based on values read)
+    3. view: list (The time-ordered list of operations performed within the Transaction to be committed to the DB server.)
+- Exceptions: 
+    1. StaleException
+    2. StaleException(Value_time: int)
+
+## Transaction Internal Structure Description
+
+First, we must understand the purpose of the Transaction's min\_Rt (min read time) and max\_Vt (max value time).  A Transaction consists of a set of updated database entries.  The database will not allow any Transaction to commit a result derived from an inconsistent snapshot of the database values.  
+
+Specifically, every entry in a snapshot of the Treode database has both read and value timestamps.  A snapshot is consistent if and only if (maximum value timestamp) < (minimum read timestamp).  (See [this explanation](https://forum.treode.com/t/eventual-consistency-and-transactions-working-together/36).)
+
+Then, by the invariant established previously that each entry is guaranteed to be the value in the DB between [V\_t, R\_t], it follows that there was a point in time when the DB contained all the values in the snapshot simultaneously.
+
+The Transaction uses min\_Rt and max\_Vt to maintain this invariant and avoid wasting network time sending commits to the DB server guaranteed to be rejected.  If any Transaction read causes (max\_Vt > min\_Rt), the read method throws a StaleException(Vt), including the value time that caused the violation.  (Note that a read time should not cause a violation because we can simply set min-age=0. Contrarily, we cannot control the value times through our read query parameters.) 
+
```

Comment:
I think we need some wordsmithing in the parenthetical note. I know what you mean, but I'm not sure others will. :-)